docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,243 @@
+  * It determines if padding is applied when comparing two strings.
+  * It describes comparisons of logically identical but physically different byte sequences.
+Some examples of the effects of collation can be found in [this part](https://dev.mysql.com/doc/refman/8.0/en/charset-collation-effect.html) of the MySQL Manual and [this webpage](http://demo.icu-project.org/icu-bin/locexp?_=en_US&x=col).
+
+Several collation implementations are described in [this part](https://dev.mysql.com/doc/refman/8.0/en/charset-collation-implementations.html) of MySQL Manual:
+  * Per-character table lookup for 8-bit character sets, for example: `latin1_swedish_ci`.
+  * Per-character table lookup for multibyte character sets, including some of the Unicode collations, for example: `utf8mb4_general_ci`.
+    - Full UCA implementation for some Unicode collations, for example: `utf8mb4_900_ai_ci`.
+
+## Proposal
+
+### Collate Interface
+
+There are two basic functions needed for collation `sortKey` calculation:
+  * Function that is used to compare two strings according to the given collation.
+  * Function that is used to make a memory-comparable `sortKey` corresponding to the given string.
+The interface can be defined as following:
+
+```
+type CollAttrFlag int
+
+const (
+    FlagIgnoreCase uint64=1 << iota
+    FlagIgnoreAccent
+    FlagPadding
+)
+
+type CharsetCollation interface {
+    // Compare returns an integer comparing the two byte slices. The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
+    Compare(a, b []byte) (int, error)
+    // CompareString returns an integer comparing the two strings. The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
+    CompareString(a, b string) (int, error)
+    // Key returns the collation key for str, the returned slice will point to an allocation in Buffer.
+    Key(buf *Buffer, str []byte) error
+    // KeyFromString returns the collation key for str, the returned slice will point to an allocation in Buffer.
+    KeyFromString(buf *Buffer, str string) error
+
+    /******* Possible Optimization ********/
+    // FromKey returns the original byte slice from collation key.
+    FromKey(buf *Buffer, str []byte) error
+    // StringFromKey returns the original string from collation key.
+    StringFromKey(str []byte) string
+}
+```
+The interface is quite similar to the Go [collate package](https://godoc.org/golang.org/x/text/collate), it is possible to implement some of the MySQL collation by simply mapping it to an existing Go collation.
+
+### Row Format
+
+The encoding layout of TiDB has been described in our [previous article](https://pingcap.com/blog/2017-07-11-tidbinternal2/#map). The row format should be changed to make it memory comparable, this is important to the index lookup. Basic principle is that all keys encoded should use the `sortKeys` result from `Key()`/`KeyFromString()` function. However, most of the `sortKeys` calculations are not reversible.
```

Comment:
> Basic principle is that all keys encoded should use the `sortKeys` result from `Key()`/`KeyFromString()` function.

Not all keys, only the key having collction.

Reply:
Addressed.